### PR TITLE
port from babylon to r3f/three

### PR DESCRIPTION
**Pull request by cr1spyvada**
- [x] Initialize  and setup @react-three/fiber in VVPlus Designer app
- [x] Port Project over to React-three-fiber from BabylonJS
- [x] Add Camera
- [x] Add Objects
- [x] Modular Code
- [x] Remove BabylonJS library dependencies
- [x] Code Clean up

### CODE REVIEW

File: apps/designer/package.json [hunk 1]
```diff
@@ -1,63 +1,57 @@
 {
-    "name": "@vvplus/web",
```
Comment (sutyum) [inline apps/designer/package.json:2]: diff the changes to this file and share as a response to the comment

File: apps/designer/public/cameraAngles.json [hunk 2]
```diff
@@ -1,106 +1,22 @@
...
-            },
-            "properties": {
-
-            }
-        }
-    }
-]
\ No newline at end of file
+  {
+    "p": [230, 25, 0],
```
Comment (sutyum) [inline apps/designer/public/cameraAngles.json]: Your JSON schema lacks metadata. atleast add name for each item here. Use descriptive names, p and t are bad names for anything but intermediate values

File: apps/designer/src/components/Accordion.tsx [hunk 3]
```diff
@@ -18,29 +18,21 @@ const Accordion = () => {
           index={1}
           activeIndex={activeIndex}
           setActiveIndex={setActiveIndex}
-          r={1000}
-          a={-9.424777961}
-          b={1.570796327}
-          t={{ x: 250, y: 70, z: 30 }}
+          p={[220, 0, 50]}
```
Comment (sutyum) [inline apps/designer/src/components/Accordion.tsx]: this is too manual. You should refactor this to an array of dictionary and then map over it to create all these Accordions

File: apps/designer/src/components/AccordionLayout.tsx [hunk 4]
```diff
@@ -1,8 +1,7 @@
 import React from "react";
-
 import { AiOutlineDown, AiOutlineRight } from "react-icons/ai";
-import animateActiveCamera from "./Animations/animateCamera";
-import { scene_variable } from "./elements/Scene";
+import useStore from "../global-stores/store";
```
Comment (sutyum) [inline apps/designer/src/components/AccordionLayout.tsx]: useStore is a bad name, you should be using a specific and descriptive name

File: apps/designer/src/components/AccordionLayout.tsx [hunk 5]
```diff
@@ -34,22 +21,16 @@ const AccordionLayout = ({
...
-        x: t.x,
-        y: t.y,
-        z: t.z,
-      },
-    });
+  const updateTarget = useStore((state) => state.updateCameraTarget);
+  const updatePosition = useStore((state) => state.updateCameraPosition);
+  const getFloor = useStore((state) => state.floor);
+
+  function animation(p: number[], t: number[]) {
```
Comment (sutyum) [inline apps/designer/src/components/AccordionLayout.tsx]: use a better name than animation

File: apps/designer/src/components/Buttons/CamButton.tsx [hunk 6]
```diff
@@ -1,6 +1,4 @@
 import React from "react";
```
Comment (sutyum) [inline apps/designer/src/components/Buttons/CamButton.tsx]: is this file needed?

File: apps/designer/src/components/Buttons/CamButton.tsx [hunk 7]
```diff
@@ -21,16 +19,16 @@ export interface ButtonProps
 
 const CamButton = ({ text, r, a, b, t }: ButtonProps) => {
   function animation() {
```
Comment (sutyum) [inline apps/designer/src/components/Buttons/CamButton.tsx]: ?

File: apps/designer/src/components/SelectionPanel.tsx [hunk 8]
```diff
@@ -4,7 +4,7 @@ import Accordion from "./Accordion";
 const SelectionPanel = () => {
```
Comment (sutyum) [inline apps/designer/src/components/SelectionPanel.tsx]: why is this needed?

File: apps/designer/src/components/FloorSlider.tsx [hunk 9]
```diff
@@ -1,15 +1,46 @@
-import React from "react";
```
Comment (sutyum) [inline apps/designer/src/components/FloorSlider.tsx:1]: FloorTab Probably

File: apps/designer/src/components/FloorSlider.tsx [hunk 10]
```diff
@@ -1,15 +1,46 @@
...
     <div className="text-white text-sm select-none   relative flex items-center justify-center  mt-2  w-[380px] font-[Roboto]">
-      <div className=" border-[1px] border-amber-100 items-center flex rounded-[20px] w-[380px]  h-8 z-[100] bg-black bg-opacity-40">
-        <div className=" bg-white bg-opacity-40  text-black w-[25%] -ml-[1px]  h-8 flex items-center justify-center rounded-l-2xl z-[100]text-black bg-bray-200">
-          Outdoor
-        </div>
-        <p className="w-[25%] flex justify-center items-center">Floor 1</p>
-        <p className="w-[25%] flex justify-center items-center">Floor 2</p>
-        <p className="w-[25%] flex justify-center items-center">Floor 3</p>
+      <div className="border-[1px] border-amber-100 items-center flex rounded-[20px] w-[380px]  h-8 z-[100] bg-black bg-opacity-40">
+        {floorLevel.map((e, idx) => (
```
Comment (sutyum) [inline apps/designer/src/components/FloorSlider.tsx]: it doesn't seem e is being using. Use _

Comment (sutyum) [reply apps/designer/src/components/FloorSlider.tsx]: _ is a reasonable name for variables that are not being used

File: apps/designer/src/components/FloorSlider.tsx [hunk 11]
```diff
@@ -1,15 +1,46 @@
...
+          <div
+            key={idx}
+            onClick={() => updateFloor(idx)}
+            className={`${
```
Comment (sutyum) [inline apps/designer/src/components/FloorSlider.tsx]: hard to follow, cumbersome. prone to off by one errors. try to refactor the core logic out into a function

File: apps/designer/src/components/elements/Camera.tsx [hunk 12]
```diff
@@ -0,0 +1,60 @@
+import { OrbitControls } from "@react-three/drei";
+import { useThree } from "@react-three/fiber";
+import { useEffect, useRef } from "react";
+import { Vector3 } from "three";
+import useStore from "../../global-stores/store";
+import gsap from "gsap";
+
+let camera_variable: THREE.Camera;
+const CameraControls = () => {
```
Comment (sutyum) [inline apps/designer/src/components/elements/Camera.tsx:9]: Add comments breaking down this function into steps:
```js
// Step 1: Initialise Three
...
// Step 2: ...
```

File: apps/designer/src/components/elements/Camera.tsx [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+  const {
+    camera,
+    gl: { domElement },
+  } = useThree();
+  camera_variable = camera;
+  camera.up = new Vector3(0, 1, 0);
```
Comment (sutyum) [inline apps/designer/src/components/elements/Camera.tsx]: every time you have magic numbers in your code -> you should consider refactoring them out into a special place to store all such constants, or magic numbers. You can do better things than that many times

File: apps/designer/src/components/elements/Camera.tsx [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+  }
+  useEffect(() => {
+    cameraAnimate();
+  }, [target]);
+
+  return (
+    <OrbitControls
+      ref={controls}
+      args={[camera, domElement]}
+      panSpeed={1}
```
Comment (sutyum) [inline apps/designer/src/components/elements/Camera.tsx]: add comments to explain these values

File: apps/designer/src/components/elements/Model.tsx [hunk 15]
```diff
@@ -0,0 +1,57 @@
+import data from "../../../public/house.json";
```
Comment (sutyum) [inline apps/designer/src/components/elements/Model.tsx:1]: your code is not well formatted, make sure to use prettier

File: apps/designer/src/components/elements/Model.tsx [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+    else if (visibleFloorIdx >= idx + 1) return true;
+    else if (visibleFloorIdx === data.floors.length - 2) return true;
+    return false;
+  }
+  return (
+    <Suspense fallback={null}>
+      {floorGroupArray.map((floor, idx) => {
+        const groupProps = {
+          position: new Vector3(0, 0, 0),
+          scale: new Vector3(10, 10, 10),
```
Comment (sutyum) [inline apps/designer/src/components/elements/Model.tsx]: comment

File: apps/designer/src/components/elements/Model.tsx [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+          visible: floorVisiblilityToggle(idx),
+        };
+        return (
+          <group key={idx} {...groupProps}>
+            {floor.map((model, index) => {
+              console.log(model, " model");
```
Comment (sutyum) [inline apps/designer/src/components/elements/Model.tsx]: comment out console logs before merging

File: apps/designer/src/components/elements/Model.tsx [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+  data.floors.map((element) => {
+    floorTracker.push(element.floorStructure.length);
+    element.floorStructure.map((e) => {
+      floorUrlArray.push(data.baseUrl + e.file);
+    });
+  });
+  const models = useGLTF(floorUrlArray);
+  const floorGroupArray: typeof models[] = [];
+  let currIdx = 0;
+  for (let i = 0; i < data.floors.length; i++) {
```
Comment (sutyum) [inline apps/designer/src/components/elements/Model.tsx]: avoid raw loops, prefer replacing this with a map or equivalent

File: apps/designer/src/components/elements/Scene.tsx [hunk 19]
```diff
@@ -1,42 +1,34 @@
...
-        onRender={onRender}
-        className={`
+    <div
+      className={`
   absolute top-0 w-full h-screen
-  ${isWelcomePanelActive === true ? "z-[1]" : "z-[-100]"}`}
-      />
+  ${isWelcomePanelActive === true ? "z-[1] visible" : "z-[-100] hidden"}`}
+    >
+      <Canvas camera={{ position: [175, 40, -80] }}>
```
Comment (sutyum) [inline apps/designer/src/components/elements/Scene.tsx]: refactor such magic constants. For instance this base position could be moved to a predefined_camera_positions variable in a constants file. This variable may be named HOME_POSITION?

File: apps/designer/src/components/elements/Skybox.tsx [hunk 20]
```diff
@@ -0,0 +1,22 @@
+import { EnvironmentMap, useCubeTexture } from "@react-three/drei";
+import { ThreeElements } from "@react-three/fiber";
+
+export const Skybox = () => {
+  const skyBoxUrl = "https://assets.vvplus.cc/misc/skybox";
+  const texture = useCubeTexture(
+    ["_px.png", "_nx.png", "_py.png", "_ny.png", "_pz.png", "_nz.png"],
+    { path: skyBoxUrl }
+  );
+  const fogProps: ThreeElements["fog"] = {
```
Comment (sutyum) [inline apps/designer/src/components/elements/Skybox.tsx:11]: add comments explaining the choice of parameter values

File: apps/designer/src/components/elements/Skybox.tsx [hunk 21]
```diff
@@ -0,0 +1,22 @@
+import { EnvironmentMap, useCubeTexture } from "@react-three/drei";
+import { ThreeElements } from "@react-three/fiber";
+
+export const Skybox = () => {
+  const skyBoxUrl = "https://assets.vvplus.cc/misc/skybox";
```
Comment (sutyum) [inline apps/designer/src/components/elements/Skybox.tsx]: should ideally be refactored out into a constants file

File: apps/designer/src/global-stores/store.ts [hunk 22]
```diff
@@ -1,16 +1,40 @@
+import { Vector3 } from "three";
```
Comment (sutyum) [inline apps/designer/src/global-stores/store.ts]: "store.ts" is an extremely unspecific name

File: apps/designer/src/global-stores/store.ts [hunk 23]
```diff
@@ -1,16 +1,40 @@
...
 import create from "zustand";
+import { OrbitControls } from "three-stdlib";
 
 // Yet to be updated with isVisible concept with use of Zod and getFile. There will be several changes made later on.
 type Store = {
```
Comment (sutyum) [inline apps/designer/src/global-stores/store.ts]: store  is an extremely unspecific name

File: apps/designer/src/global-stores/store.ts [hunk 24]
```diff
@@ -1,16 +1,40 @@
...
+  cameraPosition: Vector3;
   floor: number;
+  cameraControls: OrbitControls | null;
+  updateCameraControls: (controls: OrbitControls) => void;
   changeFloor: (val: number) => void;
+  updateCameraTarget: (newTarget: number[]) => void;
+  updateCameraPosition: (newPosition: number[]) => void;
 };
 
 const useStore = create<Store>((set) => ({
```
Comment (sutyum) [inline apps/designer/src/global-stores/store.ts]: useStore is an extremely unspecific name

File: apps/designer/src/global-stores/store.ts [hunk 25]
```diff
@@ -1,16 +1,40 @@
...
-  floor: 8,
+  cameraTarget: new Vector3(0, 70, 0),
```
Comment (sutyum) [inline apps/designer/src/global-stores/store.ts]: refactor out such global constants

File: apps/designer/src/global-stores/store.ts (hunk 22 above)
Comment (sutyum) [inline apps/designer/src/global-stores/store.ts]: Why is this in a special folder and not in [globalStore](https://github.com/TechnocultureResearch/VVplus/tree/a4390c1dc9fb8be05c8e20abdcf950d766798de1/apps/designer/src/globalStore)?

**Review by sutyum — CHANGES_REQUESTED** [review]

File: apps/designer/src/global-stores/store.ts (hunk 22 above)
Comment (cr1spyvada) [reply apps/designer/src/global-stores/store.ts]: I just made changes to the file which already had the camera props in it. This file structure was made by someone else.

File: apps/designer/package.json (hunk 1 above)
Comment (cr1spyvada) [reply apps/designer/package.json:2]: removed:

Comment (cr1spyvada) [reply apps/designer/package.json:2]: I dont understand sir how to get the diff

File: apps/designer/src/globalStore/Navigation-Store/cameraControlsStore.ts [hunk 26]
```diff
@@ -0,0 +1,43 @@
...
+  updateCameraPosition(newPosition) {
+    set(() => ({
+      cameraPosition: new Vector3(...newPosition),
+    }));
+  },
+}));
+
+export default cameraControlsStore;
+
+// fix the store to return the value of visibility in order to check what floor is to be toggled off including all features in the floor
```
Comment (sutyum) [inline apps/designer/src/globalStore/Navigation-Store/cameraControlsStore.ts]:
```
// TODO: fix the store to return the value of visibility in order to check what floor is to be toggled off including all features in the floor
```

File: apps/designer/src/components/FloorToggleTab.tsx [hunk 27]
```diff
@@ -0,0 +1,56 @@
...
+    (state) => state.updateCameraPosition
+  );
+  const changeFloor = cameraControlsStore((state) => state.changeFloor);
+
+  const updateFloor = (idx: number) => {
+    changeFloor(idx);
+    updatePosition(cameraAngles[idx].cameraPosition);
+    updateTarget(cameraAngles[idx].cameraTarget);
+  };
+  function activeFloorTab(idx: number) {
```
Comment (sutyum) [inline apps/designer/src/components/FloorToggleTab.tsx]: 1. make return type explicit
2. name should indicate that this is a conditional style

**Review by sutyum — CHANGES_REQUESTED** [review]

**Review by sutyum — APPROVED** [review]
